crates/lib/src/qpu/experimental/random.rs

```diff
@@ -0,0 +1,272 @@
+                re: sub_region_size.into(),
+                im: 0.0,
+            },
+            seed,
+        }
+    }
+}
+
+impl TryFrom<ChooseRandomRealSubRegions> for Call {
+    type Error = CallError;
+
+    fn try_from(value: ChooseRandomRealSubRegions) -> std::result::Result<Self, Self::Error> {
+        Self::try_new(
+            ChooseRandomRealSubRegions::NAME.to_string(),
+            vec![
+                UnresolvedCallArgument::Identifier(value.destination.clone()),
+                UnresolvedCallArgument::Identifier(value.source.to_string()),
+                UnresolvedCallArgument::Immediate(value.sub_region_size),
+                UnresolvedCallArgument::Identifier(value.seed.clone()),
+            ],
+        )
+    }
+}
+
+impl ExternedCall for ChooseRandomRealSubRegions {
+    type Error = Error;
+
+    const NAME: &str = "choose_random_real_sub_regions";
+
+    #[allow(clippy::doc_markdown)]
+    /// Build the signature for the `PRAGMA EXTERN choose_random_real_sub_regions`
+    /// instruction. The signature is:
+    ///
+    /// rust ignore
+    ///     "(destination : mut REAL[], source : REAL[], sub_region_size : INTEGER, seed : mut INTEGER)"
+    fn build_signature() -> Result<quil_rs::instruction::ExternSignature> {
+        use quil_rs::instruction::{ExternParameter, ExternParameterType};
+
+        let parameters = vec![
+            ExternParameter::try_new(
+                "destination".to_string(),
+                true,
+                ExternParameterType::VariableLengthVector(quil_rs::instruction::ScalarType::Real),
+            ),
+            ExternParameter::try_new(
+                "source".to_string(),
+                false,
+                ExternParameterType::VariableLengthVector(quil_rs::instruction::ScalarType::Real),
+            ),
+            ExternParameter::try_new(
+                "sub_region_size".to_string(),
+                false,
+                ExternParameterType::Scalar(quil_rs::instruction::ScalarType::Integer),
+            ),
+            ExternParameter::try_new(
+                "seed".to_string(),
+                true,
+                ExternParameterType::Scalar(quil_rs::instruction::ScalarType::Integer),
+            ),
+        ]
+        .into_iter()
+        .map(|r| r.map_err(Error::from))
+        .collect::<Result<Vec<ExternParameter>>>()?;
+        Ok(quil_rs::instruction::ExternSignature::new(None, parameters))
+    }
+}
+
+/// Hardware values are 48 bits long.
+const MAX_SEQUENCER_VALUE: u64 = 0xFFFF_FFFF_FFFF;
+
+/// Hardware multiplication currently uses the lower 16 bits of
+/// the PRNG value.
+const MAX_UNSIGNED_MULTIPLIER: u64 = 0x0000_0000_FFFF;
+
+/// A valid seed value that may be used to initialize the PRNG. Such
+/// values are in the range `[1, MAX_SEQUENCER_VALUE]` and are losslessly
+/// convertible to `f64`.
+#[derive(Debug, Clone, Copy)]
+pub struct PrngSeedValue {
+    as_u64: u64,
+    pub(super) as_f64: f64,
```

Comment:
For extra data safety, I'd keep both of these private and then use accessor methods. Otherwise, we could potentially add code that modifies `as_f64` without modifying `as_u64`